font-variant-caps: small-caps options should not apply to Japanese characters

AlexsanderDamaceno opened the pull request:
#### 48e639d6e5ccb18dd22a899c96701e7f8f9d50d4
<pre>
font-variant-caps: small-caps options should not apply to Japanese characters
<a href="https://bugs.webkit.org/show_bug.cgi?id=158522">https://bugs.webkit.org/show_bug.cgi?id=158522</a>

Reviewed by NOBODY (OOPS!).

The font-variant-caps: small-caps options should not apply to Japanese characters.
When using small-caps with Japanese text, the normal variant caps should be used for the text.

* Source/WTF/wtf/text/CharacterProperties.h:
(WTF::isJapaneseText):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advanceInternal):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48e639d6e5ccb18dd22a899c96701e7f8f9d50d4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82455 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2119 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36549 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87588 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33516 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84560 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2190 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10011 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64265 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22018 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85524 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1558 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75016 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44542 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1456 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29202 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32557 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75443 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72748 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29838 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88944 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81509 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9761 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7021 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72666 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9987 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70830 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71883 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16022 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15029 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1143 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9714 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15235 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103922 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9588 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25204 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13054 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11358 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c181f2dd790b2db76d0b5ad77d7609f33fd7e4fb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/38278

nt1m commented:
Can you add a test btw?

vitorroriz commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -314,7 +314,21 @@ void ComplexTextController::collectComplexTextRuns()
         return characters;
     }();
 
+    auto isJapanese = [](const String& str) {
+        StringView textView = str;
+        for (size_t i = 0; i < textView.length(); ++i) {
+            auto ch = textView[i];
+            if ((ch >= 0x4E00 && ch <= 0x9FFF) || (ch >= 0x30A0 && ch <= 0x30FF) || (ch >= 0x4E00 && ch <= 0x9FFF))
+                return true;
+        }
+        return false;
```
I think it is better to have such functions in a place like `Source/WTF/wtf/text/CharacterProperties.h` or `TextUtil.h`. This is specially important because you need to cover both our code paths: `ComplexTextController` and `WidthIterator` since we handle small-caps on both of them, unless we can guarantee that we will always pick the same path for such content.

vitorroriz commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -314,7 +314,21 @@ void ComplexTextController::collectComplexTextRuns()
+    };
+
     auto fontVariantCaps = m_font.fontDescription().variantCaps();
+    bool isSmall = fontVariantCaps == FontVariantCaps::Small || fontVariantCaps == FontVariantCaps::AllSmall;
+    if (isJapanese(m_run.textAsString()) && isSmall)
```
please use complete meaningful names: "isSmallCapsVariant" or something like that. Same for `isJapanese`

AlexsanderDamaceno committed:
font-variant-caps: small-caps options should not apply to Japanese characters
https://bugs.webkit.org/show_bug.cgi?id=158522

Reviewed by NOBODY (OOPS!).

The font-variant-caps: small-caps options should not apply to Japanese characters.
When using small-caps with Japanese text, the normal variant caps should be used for the text.

* Source/WTF/wtf/text/CharacterProperties.h:
(WTF::isJapaneseText):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advanceInternal):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d924b54ac4e8d7f3199a2ffa8aa445487fac4612. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/38278

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/48e639d6e5ccb18dd22a899c96701e7f8f9d50d4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/38278

fantasai reviewed (changes requested):
I think this is the wrong approach, see https://bugs.webkit.org/show_bug.cgi?id=158522#c4

Also, we need to make sure we're doing something sensible for mixed-script text... it seems like this patch is taking an all-or-nothing approach for a given string.

vitorroriz commented:
> > Can you add a test btw?
> 
> @nt1m @vitorroriz I think is hard to write a test case because we need to get the font-size related to `font-variant: normal` and `font-variant: small-caps` for compare. I haven't found a way to check this, as the computed font-size property for both `small-caps` and `normal ` cases shows the same size. Do you know another way to get this information of font-size when we have font-variant set?

If I understood it correctly you want to make sure a specific variant is not applied for content classified as Japanese. Therefore, for guarding this behavior you can simply create a -ref test in which you expect that such content will present as "Normal" even when a variant is set
